enhancements/local-storage/automatic-detection.md

```diff
@@ -62,42 +63,195 @@ and we can use that information to sort devices into storage classes based on th
+    - ##### NOTE: This step is optional for users directly creating CRs. They might already know what CRs they need for step 2.
+2. PV Creation: After the user has decided how to created PVs, they can choose to create PVs with either the `LocalVolume` CR or the `LocalVolumeSet` CR.
+
+## Design Details for `Auto discovery of local devices`
+
+API scheme for `LocalVolumeDiscoveryJob` and `LocalVolumeDiscoveryResult` CR:
+
+```go
+type LocalVolumeDiscoveryJobList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LocalVolumeDiscoveryJob `json:"items"`
+}
+
+type LocalVolumeDiscoveryJob struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              LocalVolumeDiscoveryJobSpec   `json:"spec"`
+	Status            LocalVolumeDiscoveryJobStatus `json:"status,omitempty"`
+}
+
+type LocalVolumeDiscoveryJobSpec struct {
+	// Nodes on which the automatic detection policies must run.
+	// +optional
+	NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+	// If specified tolerations is the list of toleration that is passed to the
+	// LocalVolumeDiscoveryJob Daemon
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
+
+type LocalVolumeDiscoveryJobStatus struct {
+	// Conditions is a list of conditions and their status.
+	Conditions []operatorv1.OperatorCondition `json:"conditions,omitempty"`
+	// observedGeneration is the last generation change the operator has dealt with
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+}
+
+type LocalVolumeDiscoveryResultList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LocalVolumeDiscoveryResult `json:"items"`
+}
+
+type LocalVolumeDiscoveryResult struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              LocalVolumeDiscoveryResultSpec   `json:"spec"`
+	Status            LocalVolumeDiscoveryResultStatus `json:"status,omitempty"`
+}
+
+type LocalVolumeDiscoveryResultSpec struct {
+	NodeName string `json:"nodeName"`
+}
+
+type LocalVolumeDiscoveryResultStatus struct {
+	// Conditions is a list of conditions and their status.
+	Conditions          []operatorv1.OperatorCondition `json:"conditions,omitempty"`
+	DiscoveredTimeStamp string                         `json:"discoveredTimeStamp,omitempty"`
+	// Devices contains the list of devices on which LSO
+	// is capable of creating LocalPVs
+	// The devices in this list qualify these following conditions.
+	// - it should be a non-removable device.
+	// - it should not be a read-only device.
+	// - it should not be mounted anywhere
+	// - it should not be a boot device
+	// - it should not have child partition in lsblk output
+	Devices []Device `json:"devices"`
+}
+
+type Device struct {
```

Comment:
@travisn @gnufied @jsafrane 
Looking into the old PR comments, I see there lots of discussions around what device info we want to display to the users (via LocalVolumeDiscoveryResult CR).  So creating a new thread for a final confirmation on this. So can you please confirm if current `Device` struct should suffice? 

@Travis Looking into the [comment](https://github.com/openshift/enhancements/pull/237#discussion_r392512936), should we also include devices with Filesystem in this list, that is, adding the `FSType` to this list (keeping it empty in case there is no File system on the device) ?

Reply:
My view is that discovery should return the full set of devices on the node. Then the caller can decide what they want to filter out. For example, I would expect the UI to filter out devices that are not usable for creation of PVs. But the UI also might want to show all devices on the node so the user can see total storage of the node and see what is already in use.

Reply:
> > discovery should return the full set of devices on the node
> 
> I'm not against, however, then the enhancement should document how the user can detect which devices are free to be used (i.e. no PV exists && it's not used by OS).

I'll look more into this and update the design soon.